src/os/kqueue.rs

```diff
@@ -0,0 +1,329 @@
+        Self { id, timeout }
+    }
+
+    /// Get the identifier for the timer.
+    pub fn id(self) -> usize {
+        self.id
+    }
+
+    /// Get the timeout to wait for.
+    pub fn timeout(self) -> Duration {
+        self.timeout
+    }
+}
+
+unsafe impl FilterSealed for Timer {
+    fn filter(&self, flags: FilterFlags, key: usize) -> libc::kevent {
+        // Figure out the granularity of the timer.
+        let (fflags, data) = {
+            #[cfg(not(any(target_os = "dragonfly", target_os = "netbsd", target_os = "openbsd")))]
+            {
+                let subsec_nanos = self.timeout.subsec_nanos();
+
+                match (subsec_nanos % 1_000, subsec_nanos % 1_000_000, subsec_nanos) {
+                    (_, _, 0) => (
+                        libc::NOTE_SECONDS,
+                        self.timeout.as_secs().try_into().expect("too many seconds"),
+                    ),
+                    (_, 0, _) => (
+                        // Note: 0 by default means milliseconds.
+                        0,
+                        self.timeout
+                            .as_millis()
+                            .try_into()
+                            .expect("too many milliseconds"),
+                    ),
+                    (0, _, _) => (
+                        libc::NOTE_USECONDS,
+                        self.timeout
+                            .as_micros()
+                            .try_into()
+                            .expect("too many microseconds"),
+                    ),
+                    (_, _, _) => (
+                        libc::NOTE_NSECONDS,
+                        self.timeout
+                            .as_nanos()
+                            .try_into()
+                            .expect("too many nanoseconds"),
+                    ),
+                }
+            }
+
+            #[cfg(any(target_os = "dragonfly", target_os = "netbsd", target_os = "openbsd"))]
+            {
+                // OpenBSD/Dragonfly/NetBSD only supports milliseconds.
+                // NetBSD 10 supports NOTE_SECONDS et al, once Rust drops support for
+                // NetBSD 9 we can use the same code as above.
+                // See also: https://github.com/rust-lang/libc/pull/3080
+                (
+                    0,
+                    self.timeout
+                        .as_millis()
+                        .try_into()
+                        .expect("too many milliseconds"),
+                )
+            }
+        };
+
+        #[allow(clippy::needless_update)]
+        libc::kevent {
+            ident: self.id as _,
+            filter: libc::EVFILT_TIMER,
+            flags: flags | libc::EV_RECEIPT,
+            fflags,
+            data,
+            udata: key as _,
+            ..unsafe { mem::zeroed() }
+        }
+    }
+}
+
+impl Filter for Timer {}
+
+mod __private {
+    use crate::sys::FilterFlags;
+
+    #[doc(hidden)]
+    pub trait PollerSealed {}
```

Comment:
it's probably a good idea to put this into `src/os.rs::__private`, as it should imo be used for every sealed poller extension. (and would make the pattern clear)